File: common/src/main/java/com/like/common/util/DialogFragmentUtils.java
package com.like.common.util;

import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import com.like.base.entity.Host;

import java.lang.reflect.Method;

/**
 * 操作DialogFragment的工具类
 */
public class DialogFragmentUtils {
    /**
     * 显示指定的DialogFragment，并返回此DialogFragment的实例对象
     *
     * @param host
     * @param dialogFragmentClass 对话框类
     * @param bundle              构造函数需要的参数
     * @return
     */
    public static DialogFragment showDialog(Host host, Class<?> dialogFragmentClass, Bundle bundle) {
        if (dialogFragmentClass == null || host == null) {
            return null;
        }
        DialogFragment dialogFragment = null;
        try {
            // 反射创建对应DialogFragment的实例
            Method method;
            if (bundle != null) {
                method = dialogFragmentClass.getMethod("newInstance", Bundle.class);
                dialogFragment = (DialogFragment) method.invoke(null, bundle);
            } else {
                method = dialogFragmentClass.getMethod("newInstance");
                dialogFragment = (DialogFragment) method.invoke(null);
            }
            // 先移除，再添加并显示
            FragmentManager fm = host.getFragmentManager();
            FragmentTransaction ft = fm.beginTransaction();
            Fragment preFragment = fm.findFragmentByTag(dialogFragmentClass.getSimpleName());
            if (preFragment != null) {
                ft.remove(preFragment);
            }
            if (dialogFragment != null) {
                ft.add(dialogFragment, dialogFragmentClass.getSimpleName());
                ft.commitAllowingStateLoss();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return dialogFragment;
    }

    /**
     * 隐藏指定的DialogFragment
     *
     * @param host
     * @param dialogFragmentClass 对话框类
     */
    public static void hideDialog(Host host, Class<?> dialogFragmentClass) {
        if (dialogFragmentClass == null || host == null) {
            return;
        }
        FragmentManager fm = host.getFragmentManager();
        Fragment dialog = fm.findFragmentByTag(dialogFragmentClass.getSimpleName());
        if (dialog != null && !dialog.isHidden()) {
            ((DialogFragment) dialog).dismissAllowingStateLoss();
        }
    }

}

File: common/src/main/java/com/like/common/util/DialogFragmentUtils.kt
package com.like.common.util

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentManager
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.like.base.context.BaseActivity
import com.like.base.context.BaseFragment
import com.like.base.entity.Host

/**
 * 显示DialogFragment对话框
 *
 * @param T         目标DialogFragment
 * @param args      参数列表
 */
inline fun <reified T : DialogFragment> BaseActivity.showDialogFragment(args: Bundle? = null): DialogFragment? = DialogFragmentUtils.showDialog(this, T::class.java, args)

/**
 * 显示DialogFragment对话框
 *
 * @param T         目标DialogFragment
 * @param args      参数列表
 */
inline fun <reified T : DialogFragment> BaseFragment.showDialogFragment(args: Bundle? = null): DialogFragment? = DialogFragmentUtils.showDialog(this, T::class.java, args)

/**
 * 隐藏DialogFragment对话框
 *
 * @param T         目标DialogFragment
 */
inline fun <reified T : DialogFragment> BaseActivity.hideDialogFragment() = DialogFragmentUtils.hideDialog(this, T::class.java)

/**
 * 隐藏DialogFragment对话框
 *
 * @param T         目标DialogFragment
 */
inline fun <reified T : DialogFragment> BaseFragment.hideDialogFragment() = DialogFragmentUtils.hideDialog(this, T::class.java)

object DialogFragmentUtils {
    /**
     * 显示指定的DialogFragment对话框
     *
     * @param host                  [BaseActivity]或者[BaseFragment]或者[Host]
     * @param dialogFragmentClass   目标DialogFragment
     * @param args                  参数列表
     */
    @JvmStatic
    @JvmOverloads
    fun showDialog(host: Any, dialogFragmentClass: Class<out DialogFragment>, args: Bundle? = null): DialogFragment? {
        val fm = getFragmentManager(host) ?: return null
        val dialogFragment = dialogFragmentClass.newInstance() ?: return null
        // 放入参数
        if (args != null) {
            dialogFragment.arguments = args
        }

        // 先移除，再添加并显示
        val ft = fm.beginTransaction()
        if (ft != null) {
            val preFragment = fm.findFragmentByTag(dialogFragmentClass.simpleName)
            if (preFragment != null) {
                ft.remove(preFragment)
            }
            ft.add(dialogFragment, dialogFragmentClass.simpleName)
            ft.commitAllowingStateLoss()
        }
        return dialogFragment
    }

    /**
     * 隐藏指定的DialogFragment对话框
     *
     * @param host                  [BaseActivity]或者[BaseFragment]或者[Host]
     * @param dialogFragmentClass   目标DialogFragment
     */
    @JvmStatic
    fun hideDialog(host: Any, dialogFragmentClass: Class<out DialogFragment>) {
        val fm = getFragmentManager(host) ?: return
        val dialog = fm.findFragmentByTag(dialogFragmentClass.simpleName)
        if (dialog != null && !dialog.isHidden && dialog is DialogFragment) {
            dialog.dismissAllowingStateLoss()
        }
    }

    private fun getFragmentManager(host: Any): FragmentManager? = if (host is Host) {
        host.fragmentManager
    } else if (host is BaseActivity) {
        host.supportFragmentManager
    } else if (host is BaseFragment) {
        host.fragmentManager
    } else {
        null
    }
}

abstract class BaseDialogFragment<in T : ViewDataBinding> : DialogFragment() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(DialogFragment.STYLE_NO_TITLE, android.R.style.Theme_Translucent_NoTitleBar)
        isCancelable = cancelable()
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val binding = DataBindingUtil.inflate<T>(inflater, getDialogFragmentLayoutResId(), container, false) ?: return null
        initView(binding, arguments)
        if (!cancelableOnKeyBackPressed()) {
            dialog.setOnKeyListener { dialog, keyCode, event ->
                if (keyCode == KeyEvent.KEYCODE_BACK) {
                    return@setOnKeyListener true
                }
                return@setOnKeyListener false
            }
        }
        return binding.root
    }

    fun hide() {
        this.dismissAllowingStateLoss()
    }

    /**
     * 按返回键时退出对话框
     */
    open fun cancelableOnKeyBackPressed() = true

    open fun cancelable() = true

    abstract fun getDialogFragmentLayoutResId(): Int

    abstract fun initView(binding: T, arguments: Bundle?)
}